### ISSUE TITLE
Support for Full RELRO

### ISSUE THREAD

**Issue by Blacksmoke16** [crystal-lang/crystal#11046]
I was messing around with adding a package to AUR and noticed [namcap](https://wiki.archlinux.org/title/namcap) was providing a warning: `ELF file ('usr/bin/oq') lacks FULL RELRO, check LDFLAGS`, which apparently relates to an ELF hardening feature:

* https://www.redhat.com/en/blog/hardening-elf-binaries-using-relocation-read-only-relro
* https://github.com/rust-lang/rust/blob/master/src/doc/rustc/src/exploit-mitigations.md#read-only-relocations-and-immediate-binding
* https://web.archive.org/web/20200308175234/http://tk-blog.blogspot.com/2009/02/relro-not-so-well-known-memory.html

It seems this feature is enabled by default in Rust since Oct 2017.  Probably wouldn't hurt to do the same in Crystal land?  Could reference this PR: https://github.com/rust-lang/rust/pull/43170.

**Comment by ysbaddaden** [comment]
I just got the issue in NetBSD 10.0 where RELRO got enabled.

I could compile the crystal compiler itself (from a cross compiled & linked bootstrap) without any issues, but trying to compile the std specs, I ran into _runtime_ missing symbols because of RELRO (missing xmlXYZ or _unwindXYZ symbol). It took me a while to notice that the `llvm-config` script was the one injecting `-zrelro`.

Apparently we usually support RELRO, but it seems that some external libraries may not be compatible? 

I can enforce `-Wl,-znorelro` for NetBSD for a quick fixup, but that seems overkill. Maybe the `@[Link]` annotation for lib bindings could tell when they don't support RELRO? Then we could consider to enable or disable RELRO (partial or full) when invoking the linker just like we enable PIE?

Note: it may be something entirely different. We should verify the behavior on different platforms.

**Comment by straight-shoota** [comment]
Wouldn't RELRO support be something that pkgconfig should know about? 🤔 
AFAIU this is a property of a particular lib build (the binary) and can hardly be expressed in general terms for all variants of the library (the header)?

**Comment by ysbaddaden** [comment]
RELRO is a property of the final executable, set by the linker, just like PIE. Libraries may not be compatible for different reasons. At least this is how the `pkgsrc` treats it in NetBSD: https://www.netbsd.org/docs/pkgsrc/hardening.html

**Comment by ysbaddaden** [comment]
Something _very_ weird is that I could compile a Crystal compiler with RELRO but trying to compiler `std_spec` eventually failed because of the `_Unwind_Backtrace` symbol... but the compiler is raising exceptions and it should be able to call this function, so there's something fishy.

**Comment by ysbaddaden** [comment]
Today the std specs compiled in my NetBSD 10.0 VM with `-zrelro` :upside_down_face:

**Comment by ysbaddaden** [comment]
Oh my: if `crystal` runs the `cc` command then the executable fails to run with missing symbols _but_ if I pass `--cross-compile` then copy-paste the `cc` command and run it in the shell then the executable runs with RELRO enabled :thinking:

**Comment by ysbaddaden** [comment]
Nope, I only get the above issue when I link  `std_spec` from crystal on NetBSD. Compiling something has no issues, even when I pass `--link-flags="-Wl,-zrelro,-znow"` to force full RELRO.

I tried on Ubuntu 22.04, and I can compile and run `std_spec` with the above link flags without any issues.

Expect for some corner cases on some platforms, it seems that Crystal is compatible with full RELRO :shrug:

**Comment by ysbaddaden** [comment]
Oh, the link issue on netbsd disappears if I use `--single-module`. That explains why `--cross-compile` then calling `cc` worked: both will generate a single object file!
